yt/utilities/answer_testing/testing_utilities.py

```diff
@@ -0,0 +1,576 @@
+    # to be changed
+    except TypeError:
+        if isinstance(data, dict):
+            hd = _hash_dict(data)
+        elif data is None:
+            hd = hashlib.md5(bytes(str(-1).encode("utf-8"))).hexdigest()
+        else:
+            raise
+    return hd
+
+
+def _save_result(data, outputFile):
+    r"""
+    Saves the test results to the desired answer file.
+
+    Parameters
+    ----------
+    data : dict
+        Test results to be saved.
+
+    outputFile : str
+        Name of file to save results to.
+    """
+    with open(outputFile, "a") as f:
+        yaml.dump(data, f)
+
+
+def _compare_result(new_data, old_data):
+    r"""
+    Compares the just-generated test results to those that are already
+    saved.
+
+    Parameters
+    ----------
+    new_data : dict
+        Just-generated hashed test results.
+
+    old_data : dict
+        Previously saved hashed test results.
+    """
+    for key, value in new_data.items():
+        assert value == old_data[key]
+
+
+def _save_raw_arrays(arrays, answer_file, func_name):
+    r"""
+    Saves the raw arrays produced from answer tests to a file.
+
+    The structure of `answer_file` is: each test function (e.g.,
+    test_toro1d[0-None-None-0]) forms a group. Within each group is a
+    hdf5 dataset named after the test (e.g., field_values). The value
+    stored in each dataset is the raw array corresponding to that
+    test and function.
+
+    Parameters
+    ----------
+    arrays : dict
+        Keys are the test name (e.g. field_values) and the values are
+        the actual answer arrays produced by the test.
+
+    answer_file : str
+        The name of the file to save the answers to, in hdf5 format.
+
+    func_name : str
+        The name of the function (possibly augmented by pytest with
+        test parameters) that called the test functions
+        (e.g, test_toro1d).
+    """
+    with h5py.File(answer_file, "a") as f:
+        grp = f.create_group(func_name)
+        for test_name, test_data in arrays.items():
+            # Some answer tests (e.g., grid_values, projection_values)
+            # return a dictionary, which cannot be handled by h5py
+            if isinstance(test_data, dict):
+                sub_grp = grp.create_group(test_name)
+                _parse_raw_answer_dict(test_data, sub_grp)
+            else:
+                # Some tests return None, which hdf5 can't handle, and there is
+                # no proxy, so we have to make one ourselves. Using -1
+                if test_data is None:
+                    test_data = -1
+                grp.create_dataset(test_name, data=test_data)
+
+
+def _parse_raw_answer_dict(d, h5grp):
+    """
+    Doc string.
+    """
+    for k, v in d.items():
+        if isinstance(v, dict):
+            h5_sub_grp = h5grp.create_group(k)
+            _parse_raw_answer_dict(v, h5_sub_grp)
+        else:
+            if not isinstance(k, str):
+                k = str(k)
```

Comment:
```suggestion
            k = str(k)
```

Reply:
@neutrinoceros Just want to make sure I got the change right so that this conversation can be closed.